whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+u8       type
+uint32   query_len
+[]byte   query_data
+```
+
+Currently the only supported query type is `eth_call`. This can be expanded to support other protocols.
+
+1. eth_call (query type 1)
+
+   ```go
+   u32      block_id_len
+   []byte   block_id
+   u8       num_batch_call_data
+   []byte   batch_call_data
+   ```
+
+   Calls are batched to allow specifying multiple calls against the same block. These will be done in the same batch RPC call and are easier for the requestor to verify on chain.
+
+   ```go
+   [20]byte   contract_address
+   u32        call_data_len
+   []byte     call_data
+   ```
+
+## Query Response
+
+- Off-Chain
+  ```go
+  u8         version
+  u16        sender_chain_id = 0
+  [65]byte   signature
+  u32        query_request_len
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- On-Chain [WIP] - depends on whether the request is done via VAA or not, this could be chain/emitter/sequence but that wouldn’t work with faster-than-finality
+  ```go
+  u16        sender_chain_id != 0
+  [32]byte   vaa_hash
+  []byte     query_request
+  u8         num_per_chain_responses
+  []byte     per_chain_responses
+  ```
+- Per-Chain Responses
+
+  ```go
+  u16     chain_id
+  u8      type
+  uint32  response_len
+  []byte  response
+  ```
+
+  1. eth_call (query type 1) Response Body
+
+     ```go
+     u64         block_number
+     [32]byte    block_hash
+     u64         block_time_us
+     u8          num_results
+     []byte      results
+     ```
+
+     ```go
+     u32         result_len
+     []byte      result
+     ```
+
+## REST Service
+
+### Request
+
+```ts
+export interface QueryRequest {
+  bytes: string; // As a hex string
+  signatures?: string; // As a hex string
+}
+```
+
+Note that the signature may be omitted if the user is provisioned to allow unsigned requests in the CCQ server configuration. In that case,
+the CCQ server will sign the request.
+
+#### Request Headers
+
+The requests must include the following header.
+
+- "X-API-Key" must match what is configured on the CCQ server.
+
+### Response
+
+- 200 - request reached consensus in < 1m
+
+  - Signatures is the ECDSA signature + a byte for the index of the guardian in the guardian set
+  - Signatures will appear in increasing set index order
+
+  ```ts
+  export interface QueryResponse {
+    bytes: string; // As a hex string
+    signatures?: string; // As a hex string
+  }
+  ```
+
+- 400 - bad request (malformed input or bytes)
+- 403 - [future] e.g. requestor doesn’t have an account set up
+- 500 - [future] failed to reach consensus (e.g. received 14 responses but 7 with one result and 7 with another)
+- 504 - did not reach consensus in < 1m
+
+# Rollout Considerations
+
+Testing against the testnet guardian may not give us an accurate read of gathering consensus or hitting production nodes - there’s only one guardian and it relies on some public and third-party provider nodes.
```

Comment:
I believe this is no longer the case. xLabs runs a few testnet guardians right this moment and perhaps one additional guardian is looking into hosting a testnet node?

Reply:
No, this is still a problem because all the testnet guardians share the same key, so to the CCQ server they look like a single guardian.